design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,157 @@
+   `CertificateSigningRequests` would mean the removal of the `Issuer` resource.
+
+1. As `CertificateSigningRequests` are cluster scoped, they do not provide the
+   same kind of isolation that `CertificateRequests` provide- preventing tenants
+   reading others requests, namespace specific issuer authentication (e.g. Vault
+   Kubernetes auth), namespacing cert-manager.
+
+1. `CertificateSigningRequests` do not include a duration field. This means that
+   any attempt to "proxy", or copy `CertificateRequests`, would result in a
+   regression of this feature. A feature which many users of cert-manager rely
+   on.
+
+1. cert-manager has a large number of [external
+   issuers](https://cert-manager.io/docs/configuration/external/) created by
+   the community that operate over the `CertificateRequest` resource. No longer
+   supporting these resources would require these authors to completely re-write
+   these issuers to continue to be supported by cert-manager.
+
+Supporting both resource types, but making `CertificateRequests` the "priority"
+resource (i.e. the resource created by the `Certificates` controller), means
+that we don't have any regression in the features cert-manager offers, whilst
+maintaining the same security properties.
+
+## Proposal
+
+Instead of the concept of an `IssuerRef` for `CertificateRequests`,
+`CertificateSigningRequests` have the concept of a `SignerName`. Since
+`CertificateSigningRequests` are cluster scoped resources, the signer name can
+be directly mapped to the `ClusterIssuer` resource. `ClusterIssuers` will be
+referenced in the following format:
+
+```yaml
+  signerName: cert-manager.io/${ClusterIssuer}
+```
+
+Each `CertificateSigningRequest` controller will behave in the same way as the
+existing `CertificateRequest` resource, by getting the referenced
+`ClusterIssuer`, and attempting to sign. If the `ClusterIssuer` type is not
+managed by this controller, do nothing, else sign.
+
+Each `CertificateSigningRequest` controller will automatically set the
+`CertificateSigningRequest` `RequestCondition` to `Approved` if the request is
+for their managed `ClusterIssuer` type. If the controller manages that
+`ClusterIssuer` type, but the `ClusterIssuer` doesn't exist or is not ready, the
+controller will set the condition to `Pending`. If signing fails during
+processing, it will set the `CertificateSigningRequest` condition to `Failed`.
+
+
+### Signers
+
+Some special cases for some `ClusterIssuers` that need to be addressed:
+
+- SelfSigned: Makes use of annotations, and so these annotations should also be
+    present on `CertificateSigningRequests`.
+
+- Venafi: Makes use of annotations, and so these annotations should also be
+    present on `CertificateSigningRequests`.
+
+- ACME: The ACME controller creates sub-resources (`Orders`). Since
+    `CertificateSigningRequests` are cluster scoped resources, we should create
+    `Orders` in the `Cluster Resource Namespace` (default `cert-manager`).
```

Comment:
So this is where the model for HTTP01 solving needs to change to accommodate this feature - the current way we configure Ingresses to send traffic to the HTTP01 solver pods means that we cannot easily send traffic to a pod in another namespace. If a user 'edits in place' (or runs an ingress controller per namespace), it's _not_ possible to send traffic with the current way we handle HTTP01.

There _are_ resolutions available to this, but they are going to need their own design docs altogether I think.